llvm/tools/sycl-post-link/SYCLDeviceRequirements.cpp

```diff
@@ -34,18 +36,40 @@ void llvm::getSYCLDeviceRequirements(
       {"sycl_fixed_targets", "fixed_target"},
       {"reqd_work_group_size", "reqd_work_group_size"}};
 
-  for (const auto &MD : ReqdMDs) {
+  for (const auto &[MDName, MappedName] : ReqdMDs) {
     std::set<uint32_t> Values;
-    for (const Function &F : M) {
-      if (const MDNode *MDN = F.getMetadata(MD.first)) {
+    for (const Function &F : MD.getModule()) {
+      if (const MDNode *MDN = F.getMetadata(MDName)) {
         for (size_t I = 0, E = MDN->getNumOperands(); I < E; ++I)
           Values.insert(ExtractIntegerFromMDNodeOperand(MDN, I));
       }
     }
     // We don't need the "fixed_target" property if it's empty
-    if (std::string(MD.first) == "sycl_fixed_targets" && Values.empty())
+    if (std::string(MDName) == "sycl_fixed_targets" && Values.empty())
       continue;
-    Requirements[MD.second] =
+    Requirements[MappedName] =
         std::vector<uint32_t>(Values.begin(), Values.end());
   }
+
+  // There should only be at most one function with
+  // intel_reqd_sub_group_size metadata when considering the entry
+  // points of a module, but not necessarily when considering all the
+  // functions of a module: an entry point with a
+  // intel_reqd_sub_group_size can call an ESIMD function through
+  // invoke_esimd, and that function has intel_reqd_sub_group_size=1,
+  // which is valid.
+  std::optional<uint32_t> sub_group_size;
```

Comment:
```suggestion
  std::optional<uint32_t> SubGroupSize;
```

[LLVM Coding Standard](https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly)